src/Symfony/Component/Security/Guard/Firewall/GuardAuthenticationListener.php

```diff
@@ -0,0 +1,180 @@
+use Symfony\Component\Security\Guard\GuardAuthenticatorHandler;
+use Symfony\Component\Security\Guard\Token\PreAuthenticationGuardToken;
+use Symfony\Component\Security\Core\Authentication\AuthenticationManagerInterface;
+use Symfony\Component\Security\Guard\GuardAuthenticatorInterface;
+use Psr\Log\LoggerInterface;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\Exception\AuthenticationException;
+use Symfony\Component\Security\Http\Firewall\ListenerInterface;
+use Symfony\Component\Security\Http\RememberMe\RememberMeServicesInterface;
+
+/**
+ * Authentication listener for the "guard" system.
+ *
+ * @author Ryan Weaver <weaverryan@gmail.com>
+ */
+class GuardAuthenticationListener implements ListenerInterface
+{
+    private $guardHandler;
+    private $authenticationManager;
+    private $providerKey;
+    private $guardAuthenticators;
+    private $logger;
+    private $rememberMeServices;
+
+    /**
+     * @param GuardAuthenticatorHandler      $guardHandler          The Guard handler
+     * @param AuthenticationManagerInterface $authenticationManager An AuthenticationManagerInterface instance
+     * @param string                         $providerKey           The provider (i.e. firewall) key
+     * @param GuardAuthenticatorInterface[]  $guardAuthenticators   The authenticators, with keys that match what's passed to GuardAuthenticationProvider
+     * @param LoggerInterface                $logger                A LoggerInterface instance
+     */
+    public function __construct(GuardAuthenticatorHandler $guardHandler, AuthenticationManagerInterface $authenticationManager, $providerKey, array $guardAuthenticators, LoggerInterface $logger = null)
+    {
+        if (empty($providerKey)) {
+            throw new \InvalidArgumentException('$providerKey must not be empty.');
+        }
+
+        $this->guardHandler = $guardHandler;
+        $this->authenticationManager = $authenticationManager;
+        $this->providerKey = $providerKey;
+        $this->guardAuthenticators = $guardAuthenticators;
+        $this->logger = $logger;
+    }
+
+    /**
+     * Iterates over each authenticator to see if each wants to authenticate the request.
+     *
+     * @param GetResponseEvent $event
+     */
+    public function handle(GetResponseEvent $event)
+    {
+        if (null !== $this->logger) {
+            $this->logger->info('Checking for guard authentication credentials.', array('firewall_key' => $this->providerKey, 'authenticators' => count($this->guardAuthenticators)));
+        }
+
+        foreach ($this->guardAuthenticators as $key => $guardAuthenticator) {
+            // get a key that's unique to *this* guard authenticator
+            // this MUST be the same as GuardAuthenticationProvider
+            $uniqueGuardKey = $this->providerKey.'_'.$key;
+
+            $this->executeGuardAuthenticator($uniqueGuardKey, $guardAuthenticator, $event);
```

Comment:
(It's a bit late I know, But I lacked time to review it)

IIUC the code, the last guardAuthenticator have an higher precedence than other one. 
I say that because all guardAuthenticators are execute one after an other, and all are able to 
set a new response in the event and because you never check if a response it set.

That mean if the first guardAuthenticator allow an user, the second one could reject the user.
So finally, the user will be rejected. I'm not sure that is what end developer wants.